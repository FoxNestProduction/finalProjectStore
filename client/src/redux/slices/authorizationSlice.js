import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  isUserAuthorized: false,
  token: null,
};

const authorizationSlice = createSlice({
  name: 'authorization',
  initialState,
  reducers: {
    setAuthorization(state, action) {
      state.isUserAuthorized = action.payload;
    },
    setToken: (state, action) => {
      state.token = action.payload;
    },
  },

});

export const {
  setAuthorization,
  setToken,
} = authorizationSlice.actions;

export default authorizationSlice.reducer;
