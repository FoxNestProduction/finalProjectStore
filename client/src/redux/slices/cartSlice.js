import { createSlice } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import axios from 'axios';
import { allProducts } from './productsSlice';

const initialState = {
  cartItems: [],
  isLoading: false,
  isCart: false,
};

/* eslint-disable no-param-reassign */

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addToCart(state, action) {
      // 1-ий варіант - коли нам в action.payload надходить повний об'єкт item
      if (state.cartItems.length === 0 && action.payload !== null) {
        state.cartItems.push(action.payload);
      } else {
        const index = state.cartItems.findIndex((item) => item.id === action.payload.id);
        if (index === -1) {
          state.cartItems.push(action.payload);
        } else {
          state.cartItems[index].quantity += 1;
        }
      }
      /** 2-ий варіант, коли нам приходить суто id товара потрібно звернутись
       * до однієї з двох написаних нижче функцій:
       * - deleteOrAddCartByItemId , або
       * - deleteOrAddFromCartByItemIdWithValueFromState
       * Потрібно буде протестувати, який варіант зручніше, той і використовувати
       */
    },
    setCart(state, action) {
      if (state.cartItems.length === 0) {
        state.cartItems = action.payload;
      } else {
        const uniqueFilteredProducts = action.payload.products.filter((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id !== product.id);
          const mark = matchedProduct !== undefined;
          return mark;
        });
        const notUniqueFilteredProducts = action.payload.products.filter((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id === product.id);
          const mark = matchedProduct !== undefined;
          return mark;
        }).map((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id === product.id);
          product.quantity += matchedProduct.quantity;
          return product;
        });
        state.cartItems = { ...uniqueFilteredProducts, ...notUniqueFilteredProducts };
      }
    },
    setIsLoading(state, action) {
      state.isLoading = action.payload;
    },
    setIsCart(state, action) {
      state.isCart = action.payload;
    },
    deleteFromCart(state, action) {
      // 1 - ий варіант - коли нам в action.payload надходить повний об'єкт item
      if (state.cartItems.length) {
        const index = state.cartItems.findIndex((item) => item.id === action.payload.id);

        if (index !== -1) {
          state.cartItems[index] = (
            state.cartItems[index] === 1
              ? state.cartItems.splice(index, 1)
              : state.cartItems[index].quantity -= 1
          );
        }
      }
      /** 2-ий варіант, коли нам приходить суто id товара потрібно звернутись
         * до однієї з двох написаних нижче функцій:
         * - deleteOrAddCartByItemId , або
         * - deleteOrAddFromCartByItemIdWithValueFromState
         * Потрібно буде протестувати, який варіант зручніше, той і використовувати
         */
    },
  },
});

export const { addToCart, setCart, setIsLoading, deleteFromCart, setIsCart } = cartSlice.actions;

export const getCartItemsFromServer = () => async (dispatch) => {
  try {
    dispatch(setIsLoading(true));
    const userId = useSelector((state) => state.user.user.id);
    const { data } = await axios.get('http://localhost:4000/cart');

    if (data !== null && data !== undefined) {
      const cartContainer = data.find((cartObj) => cartObj.customerId.id === userId);

      if (cartContainer !== null && data !== undefined) {
        dispatch(setIsCart(true));
        dispatch(setCart(cartContainer));
      } else {
        dispatch(setIsCart(false));
      }
    } else {
      dispatch(setIsCart(false));
    }
    // якщо data undefined, то нічого не передаємо в кошик
    // а isCart(true or false). Код далі не виконується і дані на сторінці відмальовуються
    // тільки зі стора

    // dispatch(setCart(cartContainer)); // якщо ми тут оновимо state то оновиться і localStorage
    //  тому тут потрібно написати логіку злиття даних з localStorage з даними з сервера.
    dispatch(setIsLoading(false));
  } catch (error) {
    console.warn('Error loading cart:', error);
    dispatch(setIsLoading(false));
    dispatch(setIsCart(false));
  }
};

/** 2 варіанти функції якщо ми додаємо або видаляємо товар в state cart через прокидування id:
 * TODO: варіант A:
 * ми робимо підписку на необхідний нам state всіх продуктів в самій функції.
 * Знаходимо по id необхідний нам об'єкт, та прокидуємо його в один з reducer-ов
 * за допомогою одного з ключів DELETE або ADD. Необхідний нам об'єкт падає
 * в визначений reducer та додається до кошику, або ж видаляється з нього.
 * Ця функція передається через dispatch. В неї прокидується id та key.
 * dispatch(deleteOrAddCartByItemId(id, key))
 */
export const deleteOrAddCartByItemId = (id, key) => (dispatch) => {
  const products = useSelector((state) => state.products.products);
  if (products.length !== 0) {
    const cartItem = products.find((product) => product.id === id);
    if (cartItem !== null && cartItem !== undefined) {
      if (key === 'ADD') {
        dispatch(addToCart(cartItem));
      }
      if (key === 'DELETE') {
        dispatch(deleteFromCart(cartItem));
      }
    }
  }
};

/** 2 варіанти функції якщо ми додаємо або видаляємо товар в state cart через прокидування id:
 * TODO: варіант B:
 * всередині компонента товара ми робимо підписку через useSelector на цей самий об'єкт товара.
 * Якщо точніше, то це не зовсім підписка, а пошук через функціональний селектор необхідних нам
 * даних в останній версії store, завдяки функції useSelector. В самому компоненті потрібно буде
 * прописати такий код:
 * TODO: const fullItemObject = useSelector(deleteOrAddFromCartByItemIdWithValueFromState(id))
 * в константу fullItemObject прийде повний об'єкт цього товара, і при кліку на кнопку
 * "Додати до кошика" (в нашому випадку це знак +), або видалення з кошика
 * ми передаємо в onClick такий код:
 * додати до кошика
 * TODO: onClick={dispatch(addToCart(fullItemObject))}
 * або такий, якщо видалити з кошика
 * TODO: onClick={dispatch(deleteFromCart(fullItemObject))}
 * Попередньо імпортувавши dispatch та addToCart()/deleteFromCart()
 */

export const deleteOrAddFromCartByItemIdWithValueFromState = (id) => (state) => {
  const products = allProducts(state);
  console.log(products);
  if (products.length !== 0) {
    const cartItem = products.find((product) => product.id === id);
    console.log(cartItem);
    if (cartItem !== null && cartItem !== undefined) {
      return cartItem;
    }
    return null;
  }
  return null;
};

export default cartSlice.reducer;
