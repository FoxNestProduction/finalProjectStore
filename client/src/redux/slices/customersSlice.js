import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  customers: [],
};

/* eslint-disable no-param-reassign */

const customersSlice = createSlice({
  name: 'customers',
  initialState,
  reducers: {
    setCustomers(state, action) { // eslint-disable-line no-shadow
      state.customers = action.payload;
    },
  },
});

export const getCustomers = () => async (dispatch) => {
  try {
    const { data } = await axios.get('http://localhost:4000/api/customers');
    console.log(data);
    dispatch(setCustomers(data));// eslint-disable-line no-use-before-define
  } catch (error) {
    console.log('%cError loading customers:', 'color: red; font-weight: bold;', error);
  }
};

export const { setCustomers } = customersSlice.actions;

/* eslint-enable no-param-reassign */

export default customersSlice.reducer;
