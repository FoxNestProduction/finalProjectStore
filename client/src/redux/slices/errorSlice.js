import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  error: '',
};

/* eslint-disable no-param-reassign */

const errorSlice = createSlice({
  name: 'error',
  initialState,
  reducers: {
    setError: (state, action) => {
      state.error = action.payload;
    },
  },

});

export const {
  setError,
} = errorSlice.actions;

/* eslint-enable no-param-reassign */

export default errorSlice.reducer;
