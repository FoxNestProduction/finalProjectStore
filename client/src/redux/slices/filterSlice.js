import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { instance } from '../../API/instance';
import { resetSearch, setInputSearchValue, setSearch } from './searchSlice';
import { setError, setLoading } from '../extraReducersHelpers';

export const fetchFilteredProducts = createAsyncThunk(
  'filter/fetchFilteredProducts',
  async (queryString, { rejectWithValue, dispatch }) => {
    try {
      const response = await instance.get(`/products/filter${queryString}`);
      if (response.data.products.length > 0) {
        // dispatch(setSearch([]));
        // dispatch(setInputSearchValue(''));
        dispatch(resetSearch());
      }
      console.log(response.data);
      return response.data;
    } catch (err) {
      return rejectWithValue(err.response.data);
    }
  },
);

const initialState = {
  filteredProducts: [],
  productsQuantity: null,
  loading: false,
  error: null,
  nothingFound: false,
  filterParams: {
    filterCategories: [],
    isTrending: false,
    rating: null, // mostPopular
    isHealthy: false,
    isSupreme: false,
    minPrice: 0,
    maxPrice: 30,
    sort: '',
    startPage: 1,
    perPage: 10,
  },
};

const filterSlice = createSlice({
  name: 'filter',
  initialState,
  reducers: {
    setFilteredProducts(state, action) {
      state.filteredProducts = action.payload;
    },
    setProductsQuantity(state, action) {
      state.productsQuantity = action.payload;
    },
    setNothingFound(state, action) {
      state.nothingFound = action.payload;
    },
    setFilterParams(state, action) {
      console.log('action.payload', action.payload);
      const newState = { ...state.filterParams, ...action.payload };
      console.log('newState', newState);
      state.filterParams = newState;
      // state.filterParams = { ...state.filterParams, ...action.payload };
    },
    resetFilter(state) {
      state.filteredProducts = [];
      state.productsQuantity = null;
      state.nothingFound = false;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchFilteredProducts.pending, setLoading)
      .addCase(fetchFilteredProducts.fulfilled, (state, action) => {
        if (action.payload.length === 0) {
          state.filteredProducts = [];
          state.nothingFound = true;
        } else {
          state.filteredProducts = action.payload.products;
          state.productsQuantity = action.payload.productsQuantity;
          state.nothingFound = false;
        }
        state.loading = false;
      })
      .addCase(fetchFilteredProducts.rejected, setError);
  },
});

export const { setFilteredProducts, setFilterParams,
  setNothingFound, setProductsQuantity, resetFilter } = filterSlice.actions;

export default filterSlice.reducer;
