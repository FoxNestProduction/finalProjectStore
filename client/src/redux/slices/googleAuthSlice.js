import { createSlice } from '@reduxjs/toolkit';

const initialState = {

};

const googleAuthSlice = createSlice({
  name: 'googleAuth',
  initialState,
  reducers: {
    redirect() {
      // const url = genera
    },
    setOrderInfo(state, action) {
      state.orderInfo = action.payload;
    },
    setConfirmedOrder(state, action) {
      state.confirmedOrder = action.payload;
    },
  },
});

export const {
  setOrderInfo,
  setConfirmedOrder,
} = googleAuthSlice.actions;
