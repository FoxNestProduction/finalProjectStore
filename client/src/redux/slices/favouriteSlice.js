import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';
import { useSelector } from 'react-redux';
import { setUser } from './userSlice';
import { setAuthorizationError } from './errorSlice';

const initialState = {
  favourites: [],
  cardStates: {},
};

const favouriteSlice = createSlice({
  name: 'favourites',
  initialState,
  reducers: {
    setFavourite(state, action) {
      state.favourites = action.payload;
      // localStorage.setItem('favourites', JSON.stringify(state.favourites));
      action.payload.forEach((id) => {
        state.cardStates[id] = true;
      });
    },
    addFavourite(state, action) {
      const { id } = action.payload;
      state.favourites.push(action.payload.id);
      // localStorage.setItem('favourites', JSON.stringify(state.favourites));
      state.cardStates[id] = true;
    },
    removeFavourite(state, action) {
    // eslint-disable-next-line no-underscore-dangle
      const { id } = action.payload;
      const newStateFavourites = state.favourites.filter((item) => item !== action.payload.id);
      state.favourites = newStateFavourites;
      // localStorage.setItem('favourites', JSON.stringify(newStateFavourites));
      state.cardStates[id] = false;
    },
    updateFavourites(state, action) {
      state.favourites = action.payload;
    },
  },
});

export const {
  setIsFavourite,
  addFavourite,
  removeFavourite,
  setFavourite,
  isFavourite,
  updateFavourites,
} = favouriteSlice.actions;

export default favouriteSlice.reducer;
