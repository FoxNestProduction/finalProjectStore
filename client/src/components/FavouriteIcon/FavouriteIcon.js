import React from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@mui/material';
import FavoriteIcon from '@mui/icons-material/Favorite';
import FavoriteBorderOutlinedIcon from '@mui/icons-material/FavoriteBorderOutlined';
import { addToFavourites, deleteFromFavourites, setIsFavourite, delIsFavourite } from '../../redux/slices/favouriteSlice';

const FavouriteIcon = ({ id, ishovered, isactive }) => {
  const dispatch = useDispatch();
  const isFavourite = useSelector((state) => state.favourites.cardStates[id]);
  const token = useSelector((state) => state.authorization.token);

  const toggleFavourite = () => {
    if (token && !ishovered) {
      if (isFavourite) {
        dispatch(deleteFromFavourites({ id }));
      } else {
        dispatch(setIsFavourite(id));
        dispatch(addToFavourites({ id }));
      }
    }
  };

  return (
    <IconButton onClick={() => toggleFavourite()} sx={{ m: 0, p: 0 }}>
      { !isFavourite
        ? (
          <FavoriteBorderOutlinedIcon
            ishovered={ishovered.toString()}
            isactive={isactive.toString()}
            sx={{
              color: isactive ? 'primary.main' : (ishovered ? 'text.primaryLight' : 'text.header'),
              width: '24px',
              height: '24px',
              transition: 'color 0.3s ease',
              '&:hover': { color: 'secondary.main' },
              '&:active': { color: 'secondary.hover' },
            }}
          />
        )
        : <FavoriteIcon sx={{ color: 'secondary.main', width: '24px', height: '24px' }} /> }
    </IconButton>
  );
};

FavouriteIcon.propTypes = {
  id: PropTypes.string,
  ishovered: PropTypes.bool,
  isactive: PropTypes.bool,
};

FavouriteIcon.defaultProps = {
  id: '',
  ishovered: false,
  isactive: false,
};

export default FavouriteIcon;
