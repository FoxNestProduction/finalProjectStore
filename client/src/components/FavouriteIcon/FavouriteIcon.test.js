import React from 'react';
import { screen, render, fireEvent, waitFor, act } from '@testing-library/react';
import FavouriteIcon from './FavouriteIcon';
import { setIsFavourite, addToFavourites, deleteFromFavourites } from '../../redux/slices/favouriteSlice';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));
jest.mock('../../redux/slices/favouriteSlice', () => ({
  ...jest.requireActual('../../redux/slices/favouriteSlice'),
  setIsFavourite: jest.fn(),
  addToFavourites: jest.fn(),
  deleteFromFavourites: jest.fn(),
}));
const mockDispatch = jest.spyOn(require('react-redux'), 'useDispatch');

describe('Snapshot test', () => {

  test('should FavouriteIcon render when not favourite', async () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce('token');

    const { asFragment } = render(<FavouriteIcon id="1" />);
    expect(asFragment()).toMatchSnapshot();

    const favouriteIconFalse = screen.getByTestId('FavoriteBorderOutlinedIcon');

    await act(async () => {
      fireEvent.click(favouriteIconFalse);
    });

    await waitFor(() => {
      expect(favouriteIconFalse).toBeInTheDocument();
      expect(dispatch).toHaveBeenCalledTimes(2);
      expect(addToFavourites).toHaveBeenCalledWith({ id: '1' });
      expect(setIsFavourite).toHaveBeenCalled();

    });

    jest.clearAllMocks();
  });

  test('renders FavoriteIcon when favourite', async () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce('token');

    const { asFragment } = render(<FavouriteIcon id="1" isactive={false} ishovered={false} />);
    expect(asFragment()).toMatchSnapshot();

    const favouriteIconTrue = screen.getByTestId('FavoriteIcon');

    fireEvent.click(favouriteIconTrue);

    await waitFor(() => {
      expect(favouriteIconTrue).toBeInTheDocument();
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(deleteFromFavourites).toHaveBeenCalledWith({ id: "1" });
    });

    jest.clearAllMocks();
  });

  test('styles button isHovered', () => {
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce('token');
    render(
      <FavouriteIcon id="1" ishovered={true} isactive={false} />,
    );

    const button = screen.getByTestId('FavoriteBorderOutlinedIcon');
    expect(button).toHaveStyle({
      color: 'rgb(156, 39, 176)'
    });

    jest.clearAllMocks();
  });

  test('sstyles button isActive', () => {
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce('token');
    render(
      <FavouriteIcon id="1" ishovered={false} isactive={true} />,
    );

    const button = screen.getByTestId('FavoriteBorderOutlinedIcon');
    expect(button).toHaveStyle({
      color: 'rgb(156, 39, 176)'
    });

    jest.clearAllMocks();
  });

  test('sstyles button is not authorizaite', async () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce('');
    render(
      <FavouriteIcon id="1" ishovered={false} isactive={false} />,
    );

    const favouriteIconFalse = screen.getByTestId('FavoriteBorderOutlinedIcon');

    await act(async () => {
      fireEvent.click(favouriteIconFalse);
    });

    await waitFor(() => {
      expect(dispatch).not.toHaveBeenCalled();
    });

    jest.clearAllMocks();
  });
});
