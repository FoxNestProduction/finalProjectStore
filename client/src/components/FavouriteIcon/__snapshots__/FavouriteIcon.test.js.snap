// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test renders FavoriteIcon when favourite 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-u1vl96-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i8y28s-MuiSvgIcon-root"
      data-testid="FavoriteIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`Snapshot test should FavouriteIcon render when not favourite 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-u1vl96-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1gick0k-MuiSvgIcon-root"
      data-testid="FavoriteBorderOutlinedIcon"
      focusable="false"
      isactive="false"
      ishovered="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
